src/index.tsx

```diff
@@ -0,0 +1,131 @@
+import invariant from 'invariant';
+import React, { useCallback, useContext, useState } from 'react';
+import { ReactRelayContext, commitMutation } from 'react-relay';
+import { MutationConfig, OperationBase } from 'relay-runtime';
+
+import { Omit, WithOptionalFields } from './typeHelpers';
+
+export type MutationState<T extends OperationBase> = {
+  loading: boolean;
+  data: T['response'] | null;
+  error?: Error | null;
+};
+
+export type Mutate<T extends OperationBase> = (
+  config?: Partial<MutationConfig<T>>,
+) => void;
+
+type MutationConfigWithOptionalVariables<
+  T extends OperationBase
+> = WithOptionalFields<MutationConfig<T>, 'variables'>;
+
+export function useMutation<T extends OperationBase>(
+  userConfig: MutationConfigWithOptionalVariables<T>,
+): [Mutate<T>, MutationState<T>] {
+  const [state, setState] = useState<MutationState<T>>({
+    loading: false,
+    data: null,
+    error: null,
+  });
+
+  const { environment } = useContext(ReactRelayContext);
+  const {
+    configs,
+    mutation,
+    variables,
+    uploadables,
+    onCompleted,
+    onError,
+    optimisticUpdater,
+    optimisticResponse,
+    updater,
+  } = userConfig;
+
+  const mutate = useCallback(
+    (config?: Partial<Omit<MutationConfig<T>, 'mutation'>>) => {
+      const mergedConfig = {
+        configs,
+        mutation,
+        variables,
+        uploadables,
+        onCompleted,
+        onError,
+        optimisticUpdater,
+        optimisticResponse,
+        updater,
+        ...config,
+      };
+
+      invariant(mergedConfig.variables, 'you must specify variables');
+
+      setState({
+        loading: true,
+        data: null,
+        error: null,
+      });
+
+      return new Promise((resolve, reject) => {
+        commitMutation(environment, {
+          ...mergedConfig,
+          variables: mergedConfig.variables!,
+          onCompleted: (response, error) => {
+            invariant(!error, 'mutation unexpectedly completed with error');
```

Comment:
I take it these assertions are left in for production? Seeing as the error isn’t handled any further below?

Reply:
well, because of the wonkiness with upstream error handling, i think most network layers will now throw if errors show up

as you can probably tell, we pulled this right out of our internal repo, where we do have that set up.

it wouldn't be unreasonable to add back the logic to error from `onCompleted`, though.

Reply:
are you talking about calling `onError` from `onCompleted`?

Reply:
yeah – just in case someone is using a network layer that doesn't throw when there's a graphql error. wdyt?

Reply:
To rephrase my initial thought: Is there a way that someone could have configured their build chain to compile out these assertions from their entire bundle (so including this module) and then having the situation where this error is silently ignored because the call-site is never notified of an error having occurred?

In general, though, even though most apps will throw an error in this situation, that decision should perhaps left up to the call-site. It may be useful to default to throwing, but provide an option to _not_ throw like [react-relay-network-modern](https://github.com/relay-tools/react-relay-network-modern#advanced-options-2nd-argument-after-middlewares) has?

Reply:
(Anyways, this is not meant to in any way hold up this PR, just something to consider and perhaps iterate on.)

Reply:
the dev expression plugins that i know about don't strip out invariants – they just strip out the message

that said, i don't think it hurts us much to call `onError` here. we know we won't hit it in our code, and even if we do... it doesn't matter

Reply:
@itajaja thoughts on using the error handling here?

Reply:
i'll use `onError`, makes more sense